File: backend/codehive/api/routes/github_repos.py
"""API endpoints for GitHub repo listing and cloning via ``gh`` CLI."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from codehive.api.deps import get_db
from codehive.core.project import create_project
from codehive.integrations.github.repos import (
    check_gh_status,
    clone_repo,
    list_repos,
)

github_repos_router = APIRouter(prefix="/api/github", tags=["github-repos"])


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------


class GhStatusResponse(BaseModel):
    available: bool
    authenticated: bool
    username: str | None = None
    error: str | None = None


class RepoItem(BaseModel):
    name: str
    full_name: str
    description: str | None = None
    language: str | None = None
    updated_at: str | None = None
    is_private: bool = False
    clone_url: str


class RepoListResponse(BaseModel):
    repos: list[RepoItem]
    owner: str | None = None
    total: int


class CloneRequest(BaseModel):
    repo_url: str
    destination: str
    project_name: str


class CloneResponse(BaseModel):
    project_id: str
    project_name: str
    path: str
    cloned: bool


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------


@github_repos_router.get("/status", response_model=GhStatusResponse)
async def gh_status() -> GhStatusResponse:
    """Check whether ``gh`` CLI is available and authenticated."""
    result = await check_gh_status()
    return GhStatusResponse(
        available=result.available,
        authenticated=result.authenticated,
        username=result.username,
        error=result.error,
    )


@github_repos_router.get("/repos", response_model=RepoListResponse)
async def gh_repos(
    owner: str | None = Query(None, description="GitHub username or org name"),
    search: str | None = Query(None, description="Filter repos by name"),
    limit: int = Query(100, ge=1, le=1000, description="Max repos to fetch"),
) -> RepoListResponse:
    """List GitHub repositories for the authenticated user or a given owner."""
    try:
        result = await list_repos(owner=owner, search=search, limit=limit)
    except RuntimeError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

    return RepoListResponse(
        repos=[
            RepoItem(
                name=r.name,
                full_name=r.full_name,
                description=r.description,
                language=r.language,
                updated_at=r.updated_at,
                is_private=r.is_private,
                clone_url=r.clone_url,
            )
            for r in result.repos
        ],
        owner=result.owner,
        total=result.total,
    )


@github_repos_router.post("/clone", response_model=CloneResponse)
async def gh_clone(
    body: CloneRequest,
    db: AsyncSession = Depends(get_db),
) -> CloneResponse:
    """Clone a GitHub repo and create a project."""
    try:
        cloned_path = await clone_repo(
            repo_url=body.repo_url,
            destination=body.destination,
        )
    except FileExistsError as exc:
        raise HTTPException(status_code=409, detail=str(exc))
    except ValueError as exc:
        raise HTTPException(status_code=403, detail=str(exc))
    except RuntimeError as exc:
        raise HTTPException(status_code=500, detail=str(exc))

    # Create the project record
    project = await create_project(
        db,
        name=body.project_name,
        path=cloned_path,
    )

    return CloneResponse(
        project_id=str(project.id),
        project_name=project.name,
        path=cloned_path,
        cloned=True,
    )

File: web/src/api/githubRepos.ts
import { apiClient } from "./client";

export interface GhStatusResponse {
  available: boolean;
  authenticated: boolean;
  username: string | null;
  error: string | null;
}

export interface RepoItem {
  name: string;
  full_name: string;
  description: string | null;
  language: string | null;
  updated_at: string | null;
  is_private: boolean;
  clone_url: string;
}

export interface RepoListResponse {
  repos: RepoItem[];
  owner: string | null;
  total: number;
}

export interface CloneRequest {
  repo_url: string;
  destination: string;
  project_name: string;
}

export interface CloneResponse {
  project_id: string;
  project_name: string;
  path: string;
  cloned: boolean;
}

export async function fetchGhStatus(): Promise<GhStatusResponse> {
  const response = await apiClient.get("/api/github/status");
  if (!response.ok) {
    throw new Error(`Failed to fetch GitHub status: ${response.status}`);
  }
  return response.json() as Promise<GhStatusResponse>;
}

export async function fetchGhRepos(params?: {
  owner?: string;
  search?: string;
  limit?: number;
}): Promise<RepoListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.owner) searchParams.set("owner", params.owner);
  if (params?.search) searchParams.set("search", params.search);
  if (params?.limit) searchParams.set("limit", String(params.limit));

  const qs = searchParams.toString();
  const url = `/api/github/repos${qs ? `?${qs}` : ""}`;

  const response = await apiClient.get(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch repos: ${response.status}`);
  }
  return response.json() as Promise<RepoListResponse>;
}

export async function cloneRepo(
  body: CloneRequest,
): Promise<CloneResponse> {
  const response = await apiClient.post("/api/github/clone", body);
  if (!response.ok) {
    const data = await response.json().catch(() => null);
    const detail =
      data && typeof data === "object" && "detail" in data
        ? (data as { detail: string }).detail
        : `Clone failed: ${response.status}`;
    throw new Error(detail);
  }
  return response.json() as Promise<CloneResponse>;
}
